packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -23,728 +27,771 @@ const gridListItemSizeShape = PropTypes.shape({
+				newFirstIndex -= numOfGridLines * dimensionToExtent;
+			}
+
+			if (threshold.min === -Infinity) {
+				newFirstIndex = 0;
+			} else {
+				newFirstIndex = Math.min(maxFirstIndex, newFirstIndex);
+				newFirstIndex = Math.max(0, newFirstIndex);
+			}
 
-	applyStyleToExistingNode = (index, ...rest) => {
-		const node = this.getItemNode(index);
+			this.syncThreshold(maxPos);
+			this.scrollPosition = pos;
+			this.updateMoreInfo(dataSize, pos);
 
-		if (node) {
-			this.composeStyle(node.style, ...rest);
+			if (firstIndex !== newFirstIndex) {
+				this.setState({firstIndex: newFirstIndex});
+			}
 		}
-	}
-
-	applyStyleToNewNode = (index, ...rest) => {
-		const
-			{itemRenderer, getComponentProps, data} = this.props,
-			{numOfItems} = this.state,
-			key = index % numOfItems,
-			itemElement = itemRenderer({
-				data,
-				index,
-				key
-			}),
-			style = {},
-			componentProps = getComponentProps && getComponentProps(index) || {};
-
-		this.composeStyle(style, ...rest);
-
-		this.cc[key] = React.cloneElement(itemElement, {
-			...componentProps,
-			className: classNames(css.listItem, itemElement.props.className),
-			style: {...itemElement.props.style, ...style}
-		});
-	}
-
-	applyStyleToHideNode = (index) => {
-		const key = index % this.state.numOfItems;
-		this.cc[key] = <div key={key} style={{display: 'none'}} />;
-	}
-
-	positionItems ({updateFrom, updateTo}) {
-		const
-			{dataSize} = this.props,
-			{isPrimaryDirectionVertical, dimensionToExtent, primary, secondary, scrollPosition} = this;
-		let
-			{primaryPosition, secondaryPosition} = this.getGridPosition(updateFrom),
-			hideTo = 0,
-			width, height;
-
-		primaryPosition -= scrollPosition;
-		width = (isPrimaryDirectionVertical ? secondary.itemSize : primary.itemSize) + 'px';
-		height = (isPrimaryDirectionVertical ? primary.itemSize : secondary.itemSize) + 'px';
-
-		if (updateTo > dataSize) {
-			hideTo = updateTo;
-			updateTo = dataSize;
+
+		getItemNode = (index) => {
+			const ref = this.itemContainerRef;
+
+			return ref ? ref.children[index % this.state.numOfItems] : null;
 		}
 
-		// positioning items
-		for (let i = updateFrom, j = updateFrom % dimensionToExtent; i < updateTo; i++) {
-			if (this.updateFrom === null || this.updateTo === null || this.updateFrom > i || this.updateTo <= i) {
-				this.applyStyleToNewNode(i, width, height, primaryPosition, secondaryPosition);
-			} else {
-				this.applyStyleToExistingNode(i, width, height, primaryPosition, secondaryPosition);
-			}
+		getXY = (isPrimaryDirectionVertical, primaryPosition, secondaryPosition) => (isPrimaryDirectionVertical ? {x: secondaryPosition, y: primaryPosition} : {x: primaryPosition, y: secondaryPosition})
 
-			if (++j === dimensionToExtent) {
-				secondaryPosition = 0;
-				primaryPosition += primary.gridSize;
-				j = 0;
-			} else {
-				secondaryPosition += secondary.gridSize;
+		composeStyle (style, width, height, primaryPosition, secondaryPosition) {
```

Comment:
`composeStyle` was callned in `applyStyleToNewNode`. `getXY` was called in `composeStyle`.
It would be better to merge all code in `composeStyle` and `getXY` to `applyStyleToNewNode`.

Reply:
I merged `getXY` into `composeStyle`, and leave `composeStyle` as is for better readibility.